src/Assets/File.php

```diff
@@ -639,12 +640,60 @@ protected function onBeforeWrite()
             $this->OwnerID = Member::currentUserID();
         }
 
-        // Set default name
-        if (!$this->getField('Name')) {
-            $this->Name ="new-" . strtolower($this->class);
+        $name = $this->getField('Name');
+        $title = $this->getField('Title');
+
+        // It can't be blank, default to Title
+        if (!$name) {
+            $name = $title;
+        }
+
+        // Fix illegal characters
+        $filter = FileNameFilter::create();
+        $name = $filter->filter($name);
+
+        // We might have just turned it blank, so check again.
+        if (!$name) {
+            $name = 'new-' . strtolower(basename($this->class));
         }
 
-        // Propegate changes to the AssetStore and update the DBFile field
+        // Check for duplicates when the name has changed (or is set for the first time)
+        if ($this->isChanged('Name')) {
```

Comment:
Between the above `$name = ` assignments and this line no change detection will occur. If $name was blank earlier `isChanged('Name')` will be false.